joomla/cassiopeia
# [Metismenu] [Draft] Move metismenu from template to module and integrate js into build process

Pull Request for Issues #53 , [https://github.com/joomla/cassiopeia/pull/195#issuecomment-715308105](https://github.com/joomla/cassiopeia/pull/195#issuecomment-715308105), #208 and [https://github.com/joomla/cassiopeia/pull/210#issuecomment-718710233](https://github.com/joomla/cassiopeia/pull/210#issuecomment-718710233).

Alternative to PR #221 and #224 .

### Summary of Changes

This pull request (PR) moves the alternative layout for the metismenu from the Cassiopeia template to the menu module, so the metismenu becomes available for 3rd party templates, too.

Therefore it adds a css file which is loaded by the module and provides the basic functionality. Template builders only need to add own styling like it is now for "Normal" menus, too.

In addition it changes the name of the layout to something more meaningful than "metismenu" and adds a language string for a nice layout name for the layout selection in the menu module's advanced options.

The name of the js file which already had es6 syntax is changed to extension `.es6.js`, and the file is moved to the right place in the `build/media_source` folder so that it's transpiled and minified and put at the right place in the `media` folder by the build script.

Finally, missing doc block with copyright information has been added to the js file.

### Testing Instructions

If you have development environment for Joomla 4, checkout the [development-move-metismenu-from-template-to-module](https://github.com/joomla/cassiopeia/tree/development-move-metismenu-from-template-to-module) branch, pull the latest changes, run "composer install" and "npm ci", and then make a new installation.

If you don't have that, use the following package to make a new installation: (obsolete link removed)

After that, login to backend and install "Blog Sample Data" to see the metismenu being used in frontend.

Then, still in backend, check the available layouts in the advanced options of the "Main Menu Blog" module.

Last but not least, go to frontend and verify that the metismenu works as well as before. You can also duplicate it in backend to another module position, or create another metismenu, and check in frontend that those modules still work well.

Finally, verify that in folder `media/mod_menu/js` (below the Joomla root, not the template folder) the following transpiled and minified js files are there in addition to what is there without this PR:
- menu-metismenu.es6.js
- menu-metismenu.es6.min.js
- menu-metismenu.js
- menu-metismenu.min.js

### Actual result BEFORE applying this Pull Request

Layout selection in the menu module's advanced options:

![2020-10-29_02](https://user-images.githubusercontent.com/7413183/97583428-3ca51380-19f7-11eb-914f-3f8419313f46.png)

The javascript file `templates/cassiopeia/js/mod_menu/menu-metismenu.js` to load the metismenus has es6 syntax but just a plain "js" file name extension. There is no minified file of it. The js linter will not run on that.

### Expected result AFTER applying this Pull Request

Layout selection in the menu module's advanced options:

![2020-11-03_01](https://user-images.githubusercontent.com/7413183/97963576-b05c6d00-1db7-11eb-8629-878ba5085e41.png)

The metismenu (one or more than one at a time) works as well as without this PR.

The  javascript file `build/media_source/mod_menu/js/menu-metismenu.es6.js` to load the metismenu can be checked by the linter and is transpiled and copied to the right place by the build script.

### Documentation Changes Required

The alternative layout should be described in the J4 documentation for the menu module instead of the Cassiopeia template, where it would have to be documented if this PR will not be merged.
